File: benchmark/template/programs/python/micro/mb08.py
#!/usr/bin/env python3
"""MB-08 -- Strings: build NW = 200000 words, then 5 character passes, R = 20.

Frozen workload definition: methodology/06_micro_workloads.md section 4, MB-08.
Every per-round pass is the explicit character loop the pseudocode writes: no
`upper()`, no slice reversal, no `count()`/`find()`/regex for `cnt_ab`, no
`split()` for `cnt_w`, no caching of the hashes, fresh buffers each round.
Representation, pinned for the `python` configuration by the binding table in
section 4.12(b) and not the implementer's choice: passes 1-4 operate on a
`bytearray` working buffer; pass 5 operates on a native `str` value built fresh
from that buffer inside every timed round and read with `s[i]`, never as a byte
view. Every pass is written as the explicit indexed character loop of the
pinned pseudocode.
"""

import sys
import time

NW = 200000
R = 20
Q = 1000000007
SEED = 20268917

SPACE = ord(" ")
LOWER_A = ord("a")
LOWER_B = ord("b")
LOWER_Z = ord("z")
LOWER_X = ord("x")


class Lehmer:
    """Park-Miller minimal-standard LCG: state = (48271 * state) mod 2147483647."""

    __slots__ = ("state",)

    def __init__(self, seed):
        self.state = seed

    def next_int(self):
        self.state = (48271 * self.state) % 2147483647
        return self.state


def rhash(seq, length):
    """Order-sensitive rolling hash over a sequence of character codes."""
    h = 0
    for i in range(length):
        h = (h * 131 + seq[i]) % Q
    return h


def workload():
    g = Lehmer(SEED)
    words = []
    for _ in range(NW):
        length = 4 + g.next_int() % 13
        words.append(bytes(LOWER_A + g.next_int() % 26 for _ in range(length)))
    text = bytearray(b" ".join(words))
    n = len(text)

    acc = 0
    cnt_ab = 0
    cnt_w = 0
    for r in range(R):
        p = 7 * r + 11
        if text[p] == SPACE:
            text[p] = LOWER_X
        else:
            text[p] = LOWER_A + (text[p] - LOWER_A + 1) % 26

        h1 = rhash(text, n)

        upper = bytearray(n)
        for i in range(n):
            c = text[i]
            upper[i] = c - 32 if LOWER_A <= c <= LOWER_Z else c
        h2 = rhash(upper, n)

        rev = bytearray(n)
        for i in range(n):
            rev[i] = text[n - 1 - i]
        h3 = rhash(rev, n)

        cnt_ab = 0
        for i in range(n - 1):
            if text[i] == LOWER_A and text[i + 1] == LOWER_B:
                cnt_ab = cnt_ab + 1

        # Pass 5 runs on the language's own string type, constructed fresh from
        # the working buffer inside the timed round (section 4.12(b) pins `str`
        # and `s[i]` for python); it is never a second scan of the byte buffer.
        text_str = text.decode("ascii")
        cnt_w = 1
        for i in range(n):
            if text_str[i] == " ":
                cnt_w = cnt_w + 1

        for v in (h1, h2, h3, cnt_ab, cnt_w):
            acc = (acc * 31 + v) % Q
    return acc, n, cnt_ab, cnt_w


def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "once"
    if mode == "steady":
        k_count = int(sys.argv[2]) if len(sys.argv) > 2 else 7
        for k in range(k_count):
            start = time.perf_counter_ns()
            acc, n, cnt_ab, cnt_w = workload()
            elapsed = time.perf_counter_ns() - start
            print("ITER %d %d" % (k, elapsed), flush=True)
    else:
        acc, n, cnt_ab, cnt_w = workload()
    print("MB08 %d %d %d %d" % (acc, n, cnt_ab, cnt_w))


if __name__ == "__main__":
    main()

File: benchmark/template/programs/typescript/micro/mb08.ts
// MB-08 -- Strings: text construction plus five character-level passes.
// NW = 200000 words, R = 20.
//
// Frozen workload definition: methodology/06_micro_workloads.md section 4, MB-08.
// The text is ASCII, so the byte view and the string view agree; passes 1-4
// mutate the text in place, so the working representation is a Uint8Array --
// the mutable working buffer pinned for the `typescript` configuration by the
// binding table in section 4.12(b).
// Pass 5 runs on the language's own `string` type, constructed fresh from the
// working buffer inside each timed round, and accessed with `charCodeAt(i)`,
// which is what section 4.12(b) pins for this configuration.
// Every per-round pass is the explicit character loop: no toUpperCase, no
// reverse(), no indexOf/regex for cnt_ab, no split() for cnt_w, no reuse of the
// U and V buffers between rounds, no caching of the pass-5 string.
//
// The `process` and `require` declarations are local because the frozen build
// recipe compiles a single .ts file with no @types/node package available.

declare const process: {
  argv: string[];
  hrtime: { bigint(): bigint };
};
declare function require(id: string): any;

const NW = 200000;
const R = 20;
const Q = 1000000007;
const SEED = 20268917;

/** Park-Miller / MINSTD generator: state = (48271 * state) mod 2147483647. */
class Lehmer {
  private state: number;

  constructor(seed: number) {
    this.state = seed;
  }

  nextInt(): number {
    this.state = (48271 * this.state) % 2147483647;
    return this.state;
  }
}

/** Order-sensitive rolling hash over `len` character codes. */
function rhash(seq: Uint8Array, len: number): number {
  let h = 0;
  for (let i = 0; i < len; i++) {
    h = (h * 131 + seq[i]) % 1000000007;
  }
  return h;
}

interface Mb08Result {
  acc: number;
  len: number;
  cntAb: number;
  cntW: number;
}

function workload(): Mb08Result {
  const gen = new Lehmer(SEED);

  const words: string[] = [];
  for (let w = 0; w < NW; w++) {
    const L = 4 + (gen.nextInt() % 13); // length 4..16
    const codes: number[] = new Array(L);
    for (let c = 0; c < L; c++) {
      codes[c] = 97 + (gen.nextInt() % 26); // 'a' + ...
    }
    words.push(String.fromCharCode(...codes));
  }
  const joined = words.join(" ");
  const len = joined.length;

  // Mutable ASCII view of the joined text.
  const text = new Uint8Array(len);
  for (let i = 0; i < len; i++) {
    text[i] = joined.charCodeAt(i);
  }
  const decoder = new TextDecoder();

  let acc = 0;
  let cntAb = 0;
  let cntW = 0;
  for (let r = 0; r < R; r++) {
    const p = 7 * r + 11; // anti-elimination (section 2.6)
    if (text[p] === 32) {
      text[p] = 120; // 'x'
    } else {
      text[p] = 97 + ((text[p] - 97 + 1) % 26);
    }

    const h1 = rhash(text, len); // pass 1

    const u = new Uint8Array(len); // pass 2: upper-case
    for (let i = 0; i < len; i++) {
      const c = text[i];
      u[i] = c >= 97 && c <= 122 ? c - 32 : c;
    }
    const h2 = rhash(u, len);

    const v = new Uint8Array(len); // pass 3: reverse
    for (let i = 0; i < len; i++) {
      v[i] = text[len - 1 - i];
    }
    const h3 = rhash(v, len);

    cntAb = 0; // pass 4: naive search for "ab"
    for (let i = 0; i < len - 1; i++) {
      if (text[i] === 97 && text[i + 1] === 98) {
        cntAb = cntAb + 1;
      }
    }

    // Pass 5: word count on the language's own string type. `S` is built fresh
    // from the working buffer inside the timed round and is read with the
    // pinned `charCodeAt(i)` accessor, never as a byte view (section 4.12(b)).
    const s = decoder.decode(text);
    cntW = 1;
    for (let i = 0; i < len; i++) {
      if (s.charCodeAt(i) === 32) {
        cntW = cntW + 1;
      }
    }

    for (const x of [h1, h2, h3, cntAb, cntW]) {
      acc = (acc * 31 + x) % Q;
    }
  }

  return { acc, len, cntAb, cntW };
}

function main(): void {
  // Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
  // Each steady iteration re-seeds the generator to SEED and rebuilds the words,
  // the join and the working buffer inside the timed region, so it performs
  // exactly the work `once` performs.
  const mode = process.argv.length > 2 ? process.argv[2] : "once";
  let res: Mb08Result;
  if (mode === "steady") {
    const raw = process.argv.length > 3 ? Number(process.argv[3]) : 7;
    const iterations = raw >= 1 ? Math.floor(raw) : 7;
    const fs = require("node:fs");
    res = { acc: 0, len: 0, cntAb: 0, cntW: 0 };
    for (let k = 0; k < iterations; k++) {
      const t0 = process.hrtime.bigint();
      res = workload();
      const elapsed = process.hrtime.bigint() - t0;
      // fs.writeSync(1, ..) is Node's synchronous stdout write: section 5.2
      // requires each ITER line to be flushed immediately.
      fs.writeSync(1, `ITER ${k} ${elapsed}\n`);
    }
  } else {
    res = workload();
  }

  console.log(`MB08 ${res.acc} ${res.len} ${res.cntAb} ${res.cntW}`);
}

main();
